Week 1 labs in R (#5)
prior-knowledge/bayestime: 01/01a_Central_Limit_Theorem.Rmd

---
title: 'Week 1 lab: Central Limit Theorem'
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
output_options = c("self_contained = TRUE")
```

## Goal

Use a pseudorandom number generator to test out the CLT to see that it works (except where it doesn't)

## The plan

* Pick a distribution  
* Write a function that samples from the distribution x times and returns the sample mean  
* Run the functions many times to generate a distribution of sample means- as the sample size gets bigger, does this distribution approach a normal distribution?

Let's compute means of 2, 10, 100, and 1000 samples:

```{r}
Nvals <- c(2,10,100,1000)
```

In the code below:  

> `sample_fun <- function(x) rbinom(x,1,p=0.5)`  

Reviewer: I think we're missing the `mean()`  part in the function code. Also I think calling it `sample_mean()` instead of `sample_fun()` is a bit more clear on what it is doing. Also using `n` for the argument name instead of `x` since `x` usually implies a continuous variable (just my opinion!)

Reviewer: ```suggestion
`sample_mean <- function(n) mean(rbinom(n,1,p=0.5))`  
```

> defines a function that returns a mean of x binomial samples. Since the binomial distribution can only return a 0 or 1, then sample(x=2) can only take values 0, 0.5, or 1.

Reviewer: ```suggestion
defines a function that returns a mean of x binomial samples. Since the binomial distribution can only return a 0 or 1, then sample_mean(n=2) can only take values 0, 0.5, or 1.
```

> The `for` loop runs sample_fun(x) for each value of x and plots a histogram of the sample means:

Reviewer: ```suggestion
The `for` loop runs sample_mean(n) for each value of `Nvals` and plots a histogram of the sample means:
```

> ```{r, fig.width=7}

Reviewer: Its a good idea to name code chunks, especially those that generate plots. When they are unnamed, rmarkdown produces png filenames like *unnamed-chunk-7-1.png*. These filenames do change if you add/delete/re-order chunks so it can hard to track them across versions.

Reviewer: ```suggestion
```{r plot-sample-means, fig.width=7}
```

> sample_fun <- function(x){

Reviewer: ```suggestion
sample_mean <- function(n){
```

>     mean(rbinom(x,1,p=0.5))

Reviewer: ```suggestion
    mean(rbinom(n,1,p=0.5))
```

}
par(mfrow=c(2,2))

> for(i in 1:length(Nvals)){

Reviewer: In general if we don't need the iterator counter (`i`), I think its a little cleaner to iterate over the values (`Nvals`) themselves.

Reviewer: ```suggestion
for(N in Nvals) {
```

>     N <- Nvals[i]

Reviewer: The `replicate()` function is one that people probably haven't seen before, so we just add a comment about it and pull it out from the plotting function.

Reviewer: ```suggestion
    # calculate a mean of N samples 10,000 times
    means <- replicate(10000, sample_mean(N))
```

>     hist(replicate(10000,sample_fun(N)), main=paste(N, 'sample means'), axes=FALSE, ylab="",xlab="")

Reviewer: ```suggestion
    # plot the means
    hist(means, main = paste(N, 'sample means'), axes = FALSE, ylab = "", xlab = "")
```

}
```

## What's next?

Try some different distributions:

* Poisson (generally used for models of count data- you'll see it in the next lab)  

> `sample <- function(x) mean(rpois(x,2))`  

Reviewer: ```suggestion
`sample_mean <- function(n) mean(rpois(n,2))`  
```

* Gamma (very common as a prior distribution for Poisson and Exponential models)  

> `sample <- function(x) mean(rgamma(x,7,1))`  

Reviewer: ```suggestion
`sample_mean <- function(n) mean(rgamma(n,7,1))`  
```

* Cauchy (distribution, among other things, of a ratio of Normally-distributed variables):  

> `sample <- function(x) mean(rcauchy(x))`  

Reviewer: ```suggestion
`sample_mean <- function(n) mean(rcauchy(n))`  
```
